src/modules/billing/jobs/sync-charge-categories.js

```diff
@@ -32,6 +32,14 @@ const createMessage = () => ([
   }
 ])
 
+const pick = (obj, props) => {
```

Comment:
I chose to add a function here for pick otherwise I would have had to duplicate the keys which would have been an additional 18 lines of code (since pick is used twice in this file)

Reply:
Don't worry about adding lines of code if it makes things easier to read and understand. In this case, a new team member would have to work through `pick()` to understand what it is doing. IMHO seeing two objects created from a subset of others is understandable at a glance, even if the keys are repeated.

However, in this case, I believe there is a 3rd way. Having looked at the context where `pick()` is being used I think we can refactor and make the intent more explicit.


```javascript
const handler = async () => {
  logger.info(`${JOB_NAME}: Job has started`)

  // Get the Body and the ETag of the S3 object
  const { Body, ETag } = await s3Connector.getObject(csvKey)

  // Get the current application state for the import
  const { data: currentApplicationState } = await applicationState.get('charge-categories-import')

  if (ETag === currentApplicationState.etag) {
    logger.info('No change detected. Not processing file.')
    return 'No change detected. Not processing file.'
  }

  const arraysFromCSV = await csvParse(Body, { columns: true })
  for (let i = 0; i < arraysFromCSV.length; i++) {
    const mappedChargeCategory = chargeCategoriesMapper.csvToModel(arraysFromCSV[i])
    const chargeCategoryExists = await chargeCategoriesRepo.findOneByReference(mappedChargeCategory.reference)

    if (chargeCategoryExists) {
      if (_updateExistingCategory(chargeCategoryExists, mappedChargeCategory)) {
        logger.info(`Updating an existing charge category: ${mappedChargeCategory.reference}`)
        await chargeCategoriesRepo.updateByReference(mappedChargeCategory.reference, mappedChargeCategory)
      }
    } else {
      logger.info(`Creating a new charge category: ${mappedChargeCategory.reference}`)
      await chargeCategoriesRepo.create(mappedChargeCategory)
    }
  }

  return applicationState.save('charge-categories-import', { etag: ETag })
}

const _updateExistingCategory = (existingCategory, mappedCategory) => {
  const keys = [
    'description',
    'shortDescription',
    'subsistenceCharge',
    'minVolume',
    'maxVolume',
    'isTidal',
    'lossFactor',
    'modelTier',
    'isRestrictedSource'
  ]

  for (const key of keys) {
    if (existingCategory[key] !== mappedCategory[key]) {
      return true
    }
  }

  return false
}
```

`pick()` is used to extract the values to compare to determine whether an existing charge category needs to be updated. Knowing that we can refactor things so that

- we don't depend on `lodash.pick()`
- we don't depend on `lodash.isEqual()`
- we only declare our `keys` once
- we can create a function with one purpose, and name it based on that purpose

A new member still has to look at `_updateExistingCategory()` to see what it's doing. But hopefully, the name of our function gives them a big clue. Compare that to

```javascript
const pick = (obj, props) => {
  const picked = {}
  for (const prop of props) {
    picked[prop] = obj[prop]
  }
  return picked
}

// ...

if (!isEqual(pick(chargeCategoryExists, keys), pick(mappedChargeCategory, keys))) {
 // ...
}
```

which is doing the same thing but the functionality is spread throughout the module and the meat of the logic is packed inline in the `if()`. 🤷